### ISSUE TITLE
Profiling occasionally raises error

### ISSUE THREAD

**Issue by mperham** [jhawthorn/vernier#106]
I'm seeing unpredictable but occasional errors raised from `Vernier.profile` when profiling Sidekiq jobs. The jobs retry and usually succeed the second time.

```
#<JSON::GeneratorError: source sequence is illegal/malformed utf-8>
```

I'm calling `Vernier.profile` with `out: "_filename_.json.gz"` to get gzipped JSON output. Any tips on how to further debug?

**Comment by mperham** [comment]
```
/Users/mperham/.gem/ruby/3.3.5/gems/json-2.8.1/lib/json/common.rb:318:in `generate'
/Users/mperham/.gem/ruby/3.3.5/gems/json-2.8.1/lib/json/common.rb:318:in `fast_generate'
/Users/mperham/.gem/ruby/3.3.5/gems/vernier-1.3.1/lib/vernier/output/firefox.rb:94:in `output'
/Users/mperham/.gem/ruby/3.3.5/gems/vernier-1.3.1/lib/vernier/result.rb:38:in `to_gecko'
/Users/mperham/.gem/ruby/3.3.5/gems/vernier-1.3.1/lib/vernier/result.rb:43:in `write'
/Users/mperham/.gem/ruby/3.3.5/gems/vernier-1.3.1/lib/vernier/collector.rb:114:in `stop'
/Users/mperham/.gem/ruby/3.3.5/gems/vernier-1.3.1/lib/vernier.rb:25:in `profile'
/Users/mperham/src/sidekiq/lib/sidekiq/profiler.rb:39:in `call'
/Users/mperham/src/sidekiq/lib/sidekiq/processor.rb:125:in `profile'
```

**Comment by mperham** [comment]
One thought: Sidekiq has a heartbeat background thread which runs a method called ❤️ (yes the actual emoji). Could that have an issue serializing correctly into JSON?

https://github.com/sidekiq/sidekiq/blob/0e9419a7fe5a543794d896246e0eac594ea19235/lib/sidekiq/launcher.rb#L140

**Comment by jhawthorn** [comment]
`"❤"` seems to work fine https://github.com/jhawthorn/vernier/commit/49d0b47206f0f4aa01d95df43a08d074dc07495a

Are you running the rails/active_support hook or feeding other data which is not in UTF-8? Probably the solution is running `scrub!` on whatever strings have the invalid bytes but I'd like to figure out which fields those are (or conditions they appear) rather than running it indiscriminately.

**Comment by mperham** [comment]
Nope, is there any debugging I can enable to clarify the bad data in question? I can run main if you need to add some.

**Comment by mperham** [comment]
https://github.com/ruby/json/issues/710

**Comment by mperham** [comment]
As I expected, JSON says the error is with the string `Sidekiq::Launcher#\xE2\x9D\xA4`:

```
irb(main):001> "\xE2\x9D\xA4"
=> "❤"
irb(main):002> s = _
=> "❤"
irb(main):003> s.encoding
=> #<Encoding:UTF-8>
irb(main):006> require "json"
irb(main):007> JSON.generate(s)
=> "\"❤\""
irb(main):009> JSON.generate("\xE2\x9D\xA4")
=> "\"❤\""
irb(main):010> JSON.generate("Sidekiq::Launcher#\xE2\x9D\xA4")
=> "\"Sidekiq::Launcher#❤\""
```

I'm not sure what's wrong at this point.

**Comment by jhawthorn** [comment]
My mistake 😳 this is actually failing on CI, just only on certain versions of JSON.

I'll see what I can do to fix this. It is a bit awkward with as it does seem possible, via specifying `# encoding:` or other means, to make methods with names in other encodings too.

**Comment by jhawthorn** [comment]
This should be fixed in 1.4.0

**Comment by mperham** [comment]
Confirmed, I can't reproduce it anymore! 🎉🎉🎉
